Add new meta w2v2-conformer BERT-like model

ylacombe opened the pull request:
# What does this PR do?

Meta just open-sourced a Wav2Vec2-BERT conformer [model](https://huggingface.co/facebook/w2v-bert-2.0). This one is particularly interesting because it's under a MIT license and was pretrained on 101 input languages!
 
It requires adaption to the current w2v2-conformer code, which this PR does.

cc @sanchit-gandhi, @Vaibhavs10 and @amyeroberts

ylacombe committed:
first commit

ylacombe committed:
correct default value non causal

ylacombe committed:
update config and modeling code

ylacombe committed:
update converting checkpoint

ylacombe committed:
clean modeling and fix tests

ylacombe committed:
make style

ylacombe committed:
Merge branch 'huggingface:main' into add-w2v2-shaw

ylacombe committed:
add new config parameters to docstring

ylacombe committed:
fix copied from statements

sanchit-gandhi commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -922,6 +969,121 @@ def test_inference_ctc_normal_batched_rope(self):
+        ).to(torch_device)
+        # fmt: on
+
+        torch.testing.assert_close(outputs.last_hidden_state[0, 25:35, 4:10], expected_slice_0, atol=1e-4, rtol=1e-6)
```
This is another example you can use which relies on the same premise, but uses `torch.all_close`: https://github.com/huggingface/transformers/blob/01c081d1381e20b1f5b2f31d6fa8b0af8092c0b5/tests/models/whisper/test_modeling_whisper.py#L1538

sanchit-gandhi commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -922,6 +969,121 @@ def test_inference_ctc_normal_batched_rope(self):
+
+        self.assertListEqual(list(outputs.last_hidden_state.shape), [2, 326, 1024])
+
+    def test_inference_w2v2_bert_pretrained(self):
```
WDYT about defining a function:

```python
def check_inference_pretrained(self, model_id):
```

That performs the four steps of this test:
1. Load pre-trained model + feature extractor
2. Manual masking of features
3. Forward pass of model
4. Compute cosine similarity

=> this test could then be re-used for `w2v2_bert` and also standard conformer as follows:

```python
    def test_inference_pretrained(self):
        check_inference_pretrained(self, model_id="facebook/wav2vec2-conformer-rel-pos-large")

    def test_inference_w2v2_bert_pretrained(self):
        check_inference_pretrained(self, model_id="/home/yoach/tmp/wav2vec_seamless")
```

ylacombe replied:
Great idea, I've done something similar

sanchit-gandhi reviewed (commented):
Thanks for adding this so quickly! There are a few `# Copied from` statements we can try and put back in, and also strip out any logic that we don't use in the modelling code. Otherwise LGTM

sanchit-gandhi commented on src/transformers/models/wav2vec2_conformer/convert_wav2vec2_seamless_checkpoint.py:
```diff
@@ -0,0 +1,161 @@
+    )
+    parser.add_argument("--config_path", default=None, type=str, help="Path to hf config.json of model to convert")
+    parser.add_argument("--repo_id", default=None, type=str, help="Push to this repo id if precised.")
+    parser.add_argument("--process_path", default=None, type=str, help="Push to this repo id if precised.")
```
`process_path` name could be updated / clarified, along with the "help" string here

ylacombe committed:
Apply suggestions from code review

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

ylacombe committed:
make position_embeddings_type docstrings clearer

ylacombe commented on src/transformers/models/wav2vec2_conformer/convert_wav2vec2_seamless_checkpoint.py:
```diff
@@ -0,0 +1,161 @@
+        hf_wav2vec.push_to_hub(repo_id)
+
+    if process_path:
+        processor = AutoProcessor.from_pretrained(process_path)
```
You're right, and the tokenizer is not used anyways, so I'll force `SeamlessM4TFeatureExtractor`

ylacombe committed:
clean converting script

ylacombe commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -269,6 +269,34 @@ def _sample_negative_indices(
 
 
+# Copied from transformers.models.seamless_m4t_v2.modeling_seamless_m4t_v2._compute_new_attention_mask
+def _compute_new_attention_mask(hidden_states: torch.Tensor, seq_lens: torch.Tensor):
```
nice catch

ylacombe committed:
remove function not used

ylacombe commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -544,21 +572,19 @@ def forward(self, hidden_states):
         return hidden_states, norm_hidden_states
 
 
-# Copied from transformers.models.wav2vec2.modeling_wav2vec2.Wav2Vec2FeedForward with Wav2Vec2->Wav2Vec2Conformer
```
I agree, just FYI, originally I've also added the M4T adapter (which uses attention) but the adapter is not really used currently in this new model so I removed it at the end.

ylacombe commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -856,8 +923,12 @@ def __init__(self, config):
         else:
```
Well more precisely, there is one position embedding layer per encoder layer (identified by `distance_embedding`). Because of the small context size (thanks to relative position), it doesn't take that much space

ylacombe commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -1016,18 +1091,15 @@ def forward(self, hidden_states, mask_time_indices=None):
         return codevectors, perplexity
 
 
-# Copied from transformers.models.wav2vec2.modeling_wav2vec2.Wav2Vec2Adapter with Wav2Vec2->Wav2Vec2Conformer
```
Reason explained previously! I'm putting the copied from statement back

> I agree, just FYI, originally I've also added the M4T adapter (which uses attention) but the adapter is not really used currently in this new model so I removed it at the end

ylacombe commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -1231,7 +1303,8 @@ class Wav2Vec2ConformerModel(Wav2Vec2ConformerPreTrainedModel):
         self.config = config
-        self.feature_extractor = Wav2Vec2ConformerFeatureEncoder(config)
+
+        self.feature_extractor = Wav2Vec2ConformerFeatureEncoder(config) if not config.skip_feature_encoder else None
```
This is used, I've used `config_path` in the conversion script

ylacombe committed:
clean modeling file

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28165). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ylacombe commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -210,6 +226,19 @@ def create_and_check_model_with_adapter_for_ctc(self, config, input_values, atte
             result.logits.shape, (self.batch_size, self.adapter_output_seq_length, self.vocab_size)
         )
 
+    def create_and_check_model_from_seamless_communication(self, config, input_values, attention_mask):
```
I'm using `create_and_check_causal_fbank_model` now thanks!

ylacombe committed:
apply suggestion for test file + add convert script to not_doctested

ylacombe commented:
Thanks for the many good suggestions @sanchit-gandhi, could you take one last look? 

Also cc @amyeroberts, could you review this PR? Normally I'd wait for @sanchit-gandhi's approval before asking your opinion, but as the end-of-year vacations are approaching, I'd rather speed up the process!

sanchit-gandhi commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -872,6 +919,57 @@ def _load_datasamples(self, num_samples):
 
         return [x["array"] for x in speech_samples]
 
+    def check_inference_pretrained(self, model_id, model, input_name, inputs, features_shape):
```
(nit) I believe `features_shape` can be removed as an argument since it's defined as:
```python
features_shape = inputs[input_name][:2]
```
for both fbank and raw audio models

sanchit-gandhi reviewed (approved):
LGTM - thanks for iterating @ylacombe!

amyeroberts commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -210,6 +226,19 @@ def create_and_check_model_with_adapter_for_ctc(self, config, input_values, atte
 
+    def create_and_check_causal_fbank_model(self, config, input_values, attention_mask):
+        config.add_adapter = False
+        config.skip_feature_encoder = True
+        config.skip_encoder_layer_norm = True
+        config.skip_pos_conv_embed = True
+        config.non_causal_depth_wise_conv = False
```
Is it possible to have any combination of `skip_feature_encode`, `skip_encoder_layer_norm`, `skip_pos_conv_embed` and `non_causal_depth_wise_conv` and still have a sucessful forward pass? 

It's not necessary to do all combinations, but if it is possible, one test when some are true and some are false should be added. If not possible, then there should be input validation in the config.

ylacombe replied:
Done !

amyeroberts commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -922,65 +1020,88 @@ def test_inference_ctc_normal_batched_rope(self):
+        features_shape = inputs["input_features"].shape[:2]
 
-        # ... now compare to randomly initialized model
+        # apply augmentation
+        model.wav2vec2_conformer.config.apply_spec_augment = True
```
Is this deterministic?

ylacombe replied:
It's not, but `check_inference_pretrained` sets the seed !

amyeroberts commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -922,65 +1020,88 @@ def test_inference_ctc_normal_batched_rope(self):
+        model_id = "facebook/w2v-bert-2.0"
+        model = Wav2Vec2ConformerForPreTraining.from_pretrained(model_id)
+        model.to(torch_device)
+        feature_extractor = AutoFeatureExtractor.from_pretrained(model_id)
```
Is Wav2Vec2FeatureExtractor not used for this model?

ylacombe replied:
No, this uses `SeamlessM4TFeatureExtractor` as this wav2vec model needs mel spectrograms instead of raw waveforms !

amyeroberts commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -922,65 +1020,88 @@ def test_inference_ctc_normal_batched_rope(self):
+        ).to(torch_device)
+        # fmt: on
+
+        self.assertTrue((outputs.last_hidden_state[0, 25:35, 4:10] - expected_slice_0).abs().max() <= 1e-4)
+        self.assertTrue((outputs.last_hidden_state[1, 25:35, 4:10] - expected_slice_1).abs().max() <= 1e-4)
```
Out of interest - is there any significance of these values - 25:35 and 4:10 ?

ylacombe replied:
No, it's totally arbitrary !

amyeroberts commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -872,6 +919,57 @@ def _load_datasamples(self, num_samples):
+
+        with torch.no_grad():
+            outputs_rand = model_rand(
+                inputs[input_name],
```
What's the point of passing `input_name` here - is it different for the different models?

ylacombe replied:
It is ! `input_values` for the Wav2VecFeatureExtractor and `input_features` for the `SeamlessM4TFeatureExtractor` !

amyeroberts commented on tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py:
```diff
@@ -872,6 +919,57 @@ def _load_datasamples(self, num_samples):
 
+    def check_inference_pretrained(self, model_id, model, input_name, inputs, features_shape):
+        torch.manual_seed(0)
+        mask_time_indices = _compute_mask_indices(
+            features_shape,
+            model.config.mask_time_prob,
+            model.config.mask_time_length,
+            min_masks=2,
+        )
```
It'd be better to move creating the indices to the respective model inference tests and then passing the minimal number of objects to this test: model, model config and inputs then the rest of this function becomes simpler

amyeroberts reviewed (commented):
Thanks for the work adding this model! 

Most comments are about the testing logic - it seems some code paths aren't checked and the logic can be simplified. 

Two general comments:
* Adding if/else structure to the model inits and forward passes is not advised. Would it make sense or be possible to have separate modules / model classes for the new model instead of handling this all in the existing classes? 
* From the tests, it looks like this changes the input to the model - can I still directly pass the outputs of the feature extractor as model inputs?

ylacombe commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -856,8 +897,12 @@ def __init__(self, config):
 
-        self.pos_conv_embed = Wav2Vec2ConformerPositionalConvEmbedding(config)
-        self.layer_norm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.pos_conv_embed = (
+            Wav2Vec2ConformerPositionalConvEmbedding(config) if not config.skip_pos_conv_embed else None
+        )
```
Nice catch, I believe `self.pos_conv_embed` is actually not used at all by the modeling code. 

I left it as it is, but I can totally remove it if you want!

ylacombe committed:
modify tests according to review - cleaner logic and more tests

ylacombe committed:
Apply nit suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ylacombe committed:
add checker of valid position embeddings type

ylacombe commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -591,11 +591,14 @@ def __init__(self, config):
         )
-        self.batch_norm = nn.BatchNorm1d(config.hidden_size)
+        if config.non_causal_depth_wise_conv:
+            self.batch_norm = nn.BatchNorm1d(config.hidden_size)
+        else:
+            self.depthwise_layer_norm = nn.LayerNorm(config.hidden_size)
```
There are no `batch_norm_eps` !

ylacombe committed:
instantiate new layer norm layer with the right eps

ylacombe committed:
fix freeze_feature_encoder since it can be None in some cases

ylacombe committed:
add test same output in convert script

ylacombe commented:
Hi @amyeroberts, happy new year!

I've addressed your feedback! 
Two last remarks:
1. [this commit](https://github.com/huggingface/transformers/pull/28165/commits/4848a26c1d7c8e03ba225767b26a541d3054dead) is entirely `True` and is there for training
2. this new checkpoint doesn't use `Wav2Vec2Processor` but a fbank feature extractor, here `SeamlessM4TFeatureExtractor`. Consequence -> the input name is `input_features` instead of `input_values`. Do you think we should leave this as it is ?

Let me know if you need other changes !

ArthurZucker commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -618,9 +629,18 @@ def forward(self, hidden_states):
         hidden_states = self.depthwise_conv(hidden_states)
-        hidden_states = self.batch_norm(hidden_states)
+
+        if self.non_causal_depth_wise_conv:
```
this is clearly against transformers philosophy

ArthurZucker commented on src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py:
```diff
@@ -701,6 +727,20 @@ def forward(
         else:
             scores = torch.matmul(query, key.transpose(-2, -1)) / math.sqrt(self.head_size)
 
+        if self.position_embeddings_type == "relative_key":
```
this as well

ArthurZucker reviewed (commented):
SOrry have not read the full reviews but I more pro having a new modeling file as this goes against transformers philosophy 🤗

amyeroberts commented:
+1 on @ArthurZucker's comment - @ylacombe Yes, a separate modeling file I think makes sense.

ylacombe committed:
restore wav2vec2conformer and add new model

ylacombe committed:
create processor and FE + clean

ylacombe committed:
add new model code

ylacombe committed:
fix convert script and set default config parameters

ylacombe committed:
correct model id paths

ylacombe committed:
make style

ylacombe committed:
make fix-copies and cleaning files

ylacombe committed:
fix copied from statements

ylacombe committed:
complete .md and fixe copies

ylacombe committed:
clean convert script argument defaults

ylacombe committed:
fix config parameters docstrings

ylacombe committed:
fix config docstring

ylacombe committed:
Merge branch 'huggingface:main' into add-w2v2-shaw

ylacombe committed:
add copied from and enrich FE tests

ylacombe committed:
fix copied from and repo-consistency

ylacombe committed:
add autotokenizer

ylacombe committed:
make test input length shorter and change docstring code

ylacombe committed:
fix docstrings and copied from

ylacombe committed:
add add_adapter to ASR training example

ylacombe commented:
Hey @ArthurZucker and @amyeroberts, thanks for your feedback. I've updated the codebase to add a new model instead of modifying the existing one! 
Could you review this PR again ?

Note that one failing test (torch&jax) is unrelated to my change (probably flaky) and the other one relies on a trained model checkpoint, and there are none available yet -> I'm training one at the moment and will use it if it is ready at the moment of merging. Otherwise, I'll remove/update the related docstrings

ylacombe committed:
make testing of adapters more robust

ylacombe committed:
adapt to multi adapter layers

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,2004 @@
+        hidden_states = hidden_states.view(batch_size * sequence_length * self.num_groups, -1)
+
+        if self.training:
+            # sample code vector probs via gumbel in differentiateable way
```
This is the pre-training task for Wav2Vec2 (gumbel softmax for the contrastive loss), which differs from w2v-bert: https://arxiv.org/pdf/2108.06209.pdf

=> IMO we should either remove the pre-training code and add it in a follow-up PR, or update it here to use the correct pre-training formulation

amyeroberts replied:
Was this updated for the correct formulation?

ylacombe replied:
I removed the Pretraining model, but I forgot to remove this class as well, will do now

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/processing_wav2vec2_bert.py:
```diff
@@ -0,0 +1,165 @@
+        return self.tokenizer.decode(*args, **kwargs)
+
+    @contextmanager
+    def as_target_processor(self):
```
I would remove all the `as_target_processor`/`_in_target_context_manager` logic since it's deprecated and so shouldn't be added for new models

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,2004 @@
+    )
+    def forward(
+        self,
+        input_values: Optional[torch.Tensor],
```
Here, the main model input is a tensor of fbank features (i.e. `input_features`), rather than a 1-d waveform array (i.e. `input_values`)

I would advocate for consistency with our input naming convention in the library and swapping the main model input name to `input_features`

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,2004 @@
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        extract_features = input_values
```
Redundant line: feel free to remove!

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,2004 @@
+    WAV2VEC2_BERT_START_DOCSTRING,
+)
+class Wav2Vec2BERTForXVector(Wav2Vec2BERTPreTrainedModel):
+    # Copied from transformers.models.wav2vec2_conformer.modeling_wav2vec2_conformer.Wav2Vec2ConformerForXVector.__init__ with Wav2Vec2Conformer->Wav2Vec2BERT,WAV2VEC2_CONFORMER->WAV2VEC2_BERT,wav2vec2_conformer->wav2vec2_bert
```
Every class here is copied from Wav2Vec2 conformer -> can we use a single copied from at the class level?
```python
# Copied from transformers.models.wav2vec2_conformer.modeling_wav2vec2_conformer.Wav2Vec2ConformerForXVector with ...
class Wav2Vec2BERTForXVector(Wav2Vec2BERTPreTrainedModel):
```

ylacombe replied:
This class (and the other one) removed a method as compared to the wav2vec2_conformer original class, so I had to do copied from for each method

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,2004 @@
+    WAV2VEC2_BERT_START_DOCSTRING,
+)
+class Wav2Vec2BERTForAudioFrameClassification(Wav2Vec2BERTPreTrainedModel):
+    # Copied from transformers.models.wav2vec2_conformer.modeling_wav2vec2_conformer.Wav2Vec2ConformerForAudioFrameClassification.__init__ with Wav2Vec2Conformer->Wav2Vec2BERT,WAV2VEC2_CONFORMER->WAV2VEC2_BERT,wav2vec2_conformer->wav2vec2_bert
```
Same here

ylacombe replied:
This class (and the other one) removed a method as compared to the wav2vec2_conformer original class, so I had to do copied from for each method

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,2004 @@
+
+
+@add_start_docstrings("""Wav2Vec2BERT Model with a quantizer and `VQ` head on top.""", WAV2VEC2_BERT_START_DOCSTRING)
+class Wav2Vec2BERTForPreTraining(Wav2Vec2BERTPreTrainedModel):
```
I would either update this class to use the w2v2-bert pre-training task or remove it for the first PR

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/feature_extraction_wav2vec2_bert.py:
```diff
@@ -0,0 +1,302 @@
+logger = logging.get_logger(__name__)
+
+
+# Copied from transformers.models.seamless_m4t.feature_extraction_seamless_m4t.SeamlessM4TFeatureExtractor with input_features->input_values, SeamlessM4T->Wav2Vec2BERT
```
Once we update `input_values` -> `input_features` in the modelling code, we can just use the `SeamlessM4TFeatureExtractor` directly. This FE file can then be removed in favour of using the existing class.

Let's try to then and advocate the use of the `AutoFeatureExtractor` API for loading the FE in the docs and examples, such that the user does not have to worry about loading the correct FE class: https://huggingface.slack.com/archives/C01N44FJDHT/p1667824904971239?thread_ts=1667816128.702279&cid=C01N44FJDHT

sanchit-gandhi commented on tests/models/wav2vec2_bert/test_feature_extraction_wav2vec2_bert.py:
```diff
@@ -0,0 +1,294 @@
+# coding=utf-8
```
Test file can be removed once we use the seamless FE

sanchit-gandhi commented on tests/models/wav2vec2_bert/test_modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,955 @@
+    )
+
+
+class Wav2Vec2BERTModelTester:
```
Can we add `# Copied from` statements from the Wav2Vec2 Conformer tests? Currently it's a bit difficult to pick-out which of the tests implement new logic

sanchit-gandhi commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,2004 @@
+
+    @add_start_docstrings_to_model_forward(WAV2VEC2_BERT_INPUTS_DOCSTRING)
+    @add_code_sample_docstrings(
+        checkpoint=_CHECKPOINT_FOR_DOC,
```
This checkpoint is pre-trained only, and so doesn't have a tokenizer => this means we can't run a CTC forward pass (which is why the doctest is failing: https://app.circleci.com/pipelines/github/huggingface/transformers/81743/workflows/41d2c8fa-0b3c-4eb9-88e7-3dc74f4f0c54/jobs/1050194)

Probably we can remove the docstring for the CTC class since it doesn't support this task out of the box

amyeroberts replied:
Is this just true for this checkpoint, or all checkpoints for this model?

amyeroberts replied:
For creating `input_features`, my understanding is that we only need an associated feature extractor. Could we create a custom doc example which uses a feature extractor and audio instead?

ylacombe replied:
> Is this just true for this checkpoint, or all checkpoints for this model?

[facebook/w2v-bert-2.0](https://huggingface.co/facebook/w2v-bert-2.0) is the only official checkpoint associated to the model and is only the base model. I've trained other checkpoints for CTC though, I can use them here. The other classes don't have checkpoint though. 


> For creating input_features, my understanding is that we only need an associated feature extractor. Could we create a custom doc example which uses a feature extractor and audio instead?

I'll take care of the docstrings, the official checkpoint is not up-to-date with today's change (https://huggingface.co/facebook/w2v-bert-2.0/discussions/13 and https://huggingface.co/facebook/w2v-bert-2.0/discussions/12).

sanchit-gandhi reviewed (commented):
Thanks for iterating on this! Fits much better into the Transformers design philosophy with the new modelling file as proposed offline👌

Have you double checked that the pre-training task in this PR matches the implementation in fairseq2? It currently uses the Wav2Vec2 contrastive prediction task, which differs from the w2v-bert contrastive + MLM pre-training task.

My only other main comment is that we don't need to add a new feature extractor class if we're consistent with our input namings!

Otherwise looks in good shape 👍

sanchit-gandhi commented:
Could you not convert it with `Wav2Vec2BERTModel`? It's the same as the pre-training class but without the codebook weights?

Also one point of discussion: should the model name be camel-cased as is now the convention in Transformers (`Wav2Vec2BertModel`)? This is a bit of an edge case since BERT was original named as `BERTModel` (upper-case), so it's a question of whether we want consistency with BERT, or want to adhere to the camel-case conventions.

ylacombe commented:
> Could you not convert it with Wav2Vec2BERTModel? It's the same as the pre-training class but without the codebook weights?

Every downstream model (`Wav2Vec2BERTForPreTraining`, `Wav2Vec2BERTForCTC`, `Wav2Vec2BERTForSequenceClassification`, etc.) actually has `Wav2Vec2BertModel` as a parameter (`self.wav2vec2_bert`). 

So I could convert the weights to `Wav2Vec2BERTModel`, but actually using `CLASS.from_pretrained(REPO_ID)` won't work, right ?


> Also one point of discussion: should the model name be camel-cased as is now the convention in Transformers (Wav2Vec2BertModel)? This is a bit of an edge case since BERT was original named as BERTModel (upper-case), so it's a question of whether we want consistency with BERT, or want to adhere to the camel-case conventions.

I'll let @ArthurZucker or @amyeroberts decide on this one !

sanchit-gandhi commented:
This should be handled by `base_model_prefix`:
```python
from transformers import Wav2Vec2BERTConfig, Wav2Vec2BERTModel, Wav2Vec2BERTForCTC
import torch

config = Wav2Vec2BERTConfig()
random_model = Wav2Vec2BERTModel(config)

random_model.save_pretrained("./output_dir")
loaded_model = Wav2Vec2BERTForCTC.from_pretrained("./output_dir", vocab_size=32)

# check a layer-norm layer is loaded correctly
print(torch.allclose(loaded_model.wav2vec2_bert.feature_projection.layer_norm.weight, random_model.feature_projection.layer_norm.weight))
```

ylacombe commented on src/transformers/models/wav2vec2_bert/configuration_wav2vec2_bert.py:
```diff
@@ -0,0 +1,343 @@
+            Kernel size of the convolutional layers in the adapter network. Only relevant if `add_adapter is True`.
+        adapter_stride (`int`, *optional*, defaults to 2):
+            Stride of the convolutional layers in the adapter network. Only relevant if `add_adapter is True`.
+        num_adapter_layers (`int`, *optional*, defaults to 1):
```
> Is this what fairseq2 does for w2v-bert?

This is what M4T does when using w2v-bert. 

Additionally, the previous W2V2-conformer model used a learned feature extractor, whereas here we use a mel-spectrogram representation instead. 

Given that the base model doesn't modify the input "sequence length", the output sequence length before the adapter layer typically differs between W2V2-conformer and w2v-bert, all else being equal. 

W2V2-conformer is longer than w2v-bert one, thus needing more adapter layers to reach the same ratio of input/output length.

ylacombe committed:
refactor input_values->input_features and remove w2v2-bert feature extractor

ylacombe committed:
remove pretraining model

ylacombe committed:
remove depreciated features and useless lines

ylacombe committed:
add copied from and ignore statements to modeling tests

ylacombe committed:
remove pretraining model #2

ylacombe committed:
change import in convert script

ylacombe committed:
change default in convert script

ylacombe committed:
update readme and remove useless line

ylacombe commented on tests/models/wav2vec2_bert/test_modeling_wav2vec2_bert.py:
Added Copied from statements ! You can pinpoint change with `# Ignore copy` statements!

sanchit-gandhi replied:
Looks good!

ylacombe reviewed (commented):
Thanks @sanchit-gandhi for the review! I've updated as requested.

@ArthurZucker, could you review when you have time



For future reference, the two following commits are the ones that remove Wav2Vec2BERTForPretraining: [1](https://github.com/huggingface/transformers/pull/28165/commits/a17d2fa77fe2f5e34086ab2662c377185f2bb86b) and [2](https://github.com/huggingface/transformers/pull/28165/commits/8a838314b5aecb045213c4afea4986bedb24b1df).

amyeroberts commented on tests/models/wav2vec2_bert/test_processor_wav2vec2_bert.py:
```diff
@@ -0,0 +1,155 @@
+from ..wav2vec2.test_feature_extraction_wav2vec2 import floats_list
+
+
+class Wav2Vec2BERTProcessorTest(unittest.TestCase):
```
Any test methods here which are the same for wav2vec2  processor should have copied from statements here

amyeroberts commented on tests/models/wav2vec2_bert/test_modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,919 @@
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.check_labels_out_of_vocab(*config_and_inputs)
+
+    # Wav2Vec2BERT has no inputs_embeds
+    def test_inputs_embeds(self):
+        pass
```
All skipped tests should be skipped with an explicit `@unittest.skip(reason=...)` decorator

amyeroberts commented on src/transformers/models/wav2vec2_bert/processing_wav2vec2_bert.py:
```diff
@@ -0,0 +1,136 @@
+            inputs["labels"] = encodings["input_ids"]
+            return inputs
+
+    def pad(self, *args, **kwargs):
+        """
+        This method forwards the `input_features` and `kwargs` arguments to SeamlessM4TFeatureExtractor's [`~SeamlessM4TFeatureExtractor.pad`] if `input_features` is not
+        `None` to pad the input features. To pad the label(s), this method forwards the `labels` and `kwargs` arguments to
+        PreTrainedTokenizer's [`~PreTrainedTokenizer.pad`] if `labels` is not `None`. Please refer to the doctsring of the above two methods for more information.
+        """
+        input_features = kwargs.pop("input_features", None)
+        labels = kwargs.pop("labels", None)
+        if len(args) > 0:
+            input_features = args[0]
+            args = args[1:]
```
If I call `processor.pad(a, b, input_features=(foo, bar))` this will set `input_features` to `a` even if I passed it in explicitly as a kwarg

amyeroberts replied:
Let's just define and pass in kwargs explicitly. It's clearer for the user and avoids any tricky logic.

```suggestion
    def pad(self, input_features=None, labels=None, **kwargs):
        """
        This method forwards the `input_features` and `kwargs` arguments to SeamlessM4TFeatureExtractor's [`~SeamlessM4TFeatureExtractor.pad`] if `input_features` is not
        `None` to pad the input features. To pad the label(s), this method forwards the `labels` and `kwargs` arguments to
        PreTrainedTokenizer's [`~PreTrainedTokenizer.pad`] if `labels` is not `None`. Please refer to the doctsring of the above two methods for more information.
        """
```

amyeroberts commented on src/transformers/models/wav2vec2_bert/processing_wav2vec2_bert.py:
```diff
@@ -0,0 +1,136 @@
+
+    def pad(self, *args, **kwargs):
+        """
+        This method forwards the `input_features` and `kwargs` arguments to SeamlessM4TFeatureExtractor's [`~SeamlessM4TFeatureExtractor.pad`] if `input_features` is not
+        `None` to pad the input features. To pad the label(s), this method forwards the `labels` and `kwargs` arguments to
+        PreTrainedTokenizer's [`~PreTrainedTokenizer.pad`] if `labels` is not `None`. Please refer to the doctsring of the above two methods for more information.
```
Could you rework this docstring? It's hard to understand. Having `input_features` and `labels` defined as kwargs and then these described as input arguments would help. 

In particular, it would be good to clarify what happens under the different cases e.g. when both `inpute_features` and `labels` are passed in

amyeroberts commented on src/transformers/models/wav2vec2_bert/processing_wav2vec2_bert.py:
```diff
@@ -0,0 +1,136 @@
+            args = args[1:]
+
+        if input_features is not None:
+            input_features = self.feature_extractor.pad(input_features, *args, **kwargs)
+        if labels is not None:
+            labels = self.tokenizer.pad(labels, **kwargs)
```
This is tricky - if both `input_features` and `labels` are passed in, then `**kwargs` will be passed to both the `feature_extractor` and `tokenizer` even though they likely don't accept all the same input arguments

ylacombe replied:
Let's make `text` and `audio` mutually exclusive then (this is what's done with M4T)! Users will have better control

amyeroberts commented on src/transformers/models/wav2vec2_bert/processing_wav2vec2_bert.py:
```diff
@@ -0,0 +1,136 @@
+        `None` to pre-process the audio. To prepare the target sequences(s), this method forwards the `text` and `kwargs` arguments to
+        PreTrainedTokenizer's [`~PreTrainedTokenizer.__call__`] if `text` is not `None`. Please refer to the doctsring of the above two methods for more information.
+        """
+        if "raw_speech" in kwargs:
+            warnings.warn("Using `raw_speech` as a keyword argument is deprecated. Use `audio` instead.")
+            audio = kwargs.pop("raw_speech")
+        else:
+            audio = kwargs.pop("audio", None)
+        sampling_rate = kwargs.pop("sampling_rate", None)
+        text = kwargs.pop("text", None)
+        if len(args) > 0:
+            audio = args[0]
+            args = args[1:]
```
Same comment here - lets define explicit input kwargs

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+        self.stride = config.adapter_stride
+
+        # 1. residual convolution
+        self.residual_layer_norm = nn.LayerNorm(embed_dim)
```
All layer norms layers should use the config eps to initialize itself 

```suggestion
        self.residual_layer_norm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
```

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+        self.activation = nn.GLU(dim=1)
+
+        # Self-Attention
+        self.self_attn_layer_norm = nn.LayerNorm(embed_dim)
```
Same here re eps

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+
+        # Feed-forward
+        self.ffn_layer_norm = nn.LayerNorm(embed_dim)
+        self.ffn = Wav2Vec2BERTFeedForward(config, act_fn="relu", hidden_size=embed_dim)
```
Is there a reason for allowing the activation function to be set using the config sometimes but hard-coded here e.g. does it only ever make sense to use relu here?

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+            sub_sampled_lengths = (attention_mask.size(1) - (1 - attention_mask.int()).sum(1)).to(hidden_states.device)
+
+        for layer in self.layers:
+            layerdrop_prob = np.random.random()
```
I don't think we should be using numpy's RNG here. If someone tries to make something deterministic by setting torch's seed the user is still going to get unexplained randomness

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+        # feature dim might need to be down-projected
+        if config.output_hidden_size != config.hidden_size:
+            self.proj = nn.Linear(config.hidden_size, config.output_hidden_size)
+            self.proj_layer_norm = nn.LayerNorm(config.output_hidden_size)
```
Should initialize with eps here

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+        else:
+            relative_position_embeddings = None
+
+        deepspeed_zero3_is_enabled = is_deepspeed_zero3_enabled()
```
Is this needed? It seems to be a logic control only present in audio models

amyeroberts replied:
Maybe a better question is why is this only present for audio models?

ylacombe replied:
This is a great question, I've tried to trace it back, and I've found these pointers. It was introduced for W2V2 by @stas00 in #12076, following the work in #11638 and in response to #11446.

I'm not sure how much usage Deepspeed Zero3 is getting, but since not so many models have this logic control, I would say we can remove it here. WDYT ?

amyeroberts replied:
Let's keep it here for this model and consistency, and then we can have a follow-up PR which can remove it for this + other models. This means we can isolate testing and debugging any hidden issues it might be resolving

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+        dropout = config.attention_dropout
+
+        # Feed-forward 1
+        self.ffn1_layer_norm = nn.LayerNorm(embed_dim)
```
should use `eps=` here and for layers below

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+        self.ffn2 = Wav2Vec2BERTFeedForward(config)
+        self.final_layer_norm = nn.LayerNorm(embed_dim)
+
+    # Ignore copy
+    def forward(
```
If the forward is different and the only other method is `__init__` let's just define a new class without using `# Copied from`

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+        super().__init__()
+        if (config.conv_depthwise_kernel_size - 1) % 2 == 1:
+            raise ValueError("`config.conv_depthwise_kernel_size` should be a odd number for 'SAME' padding")
+        self.layer_norm = nn.LayerNorm(config.hidden_size)
```
same here re eps

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+            bias=False,
+        )
+
+        self.depthwise_layer_norm = nn.LayerNorm(config.hidden_size)
```
and here

amyeroberts commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1749 @@
+
+        return hidden_states, probs
+
+    def _apply_rotary_embedding(self, hidden_states, relative_position_embeddings):
```
Are the rotary and relative embedding methods different from those in conformer? Could we add copied from statements here?

amyeroberts reviewed (commented):
Thanks for splitting this out into a new model! 

A few comments small-ish comments. The main ones being explicitly defining kwargs in the processor methods instead of clever logic to handle args and renaming `BERT` -> `Bert` for consistency with the rest of the library

ylacombe committed:
Update tests/models/wav2vec2_bert/test_processor_wav2vec2_bert.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ylacombe committed:
refactor BERT to Bert for consistency

ylacombe committed:
remove useless ignore copy statement

ylacombe committed:
add persistent to buffer in rotary

ylacombe committed:
add eps in LayerNorm init and remove copied from

ylacombe committed:
add adapter activation parameters and add copied from statements

ylacombe committed:
Fix copied statements and add unitest.skip reasons

ylacombe committed:
add copied statement in test_processor

ylacombe committed:
refactor processor

ylacombe committed:
make style

ylacombe committed:
replace numpy random by torch rand

ylacombe committed:
remove expected output CTC

ylacombe commented:
Hey @amyeroberts, thanks for the review ! I've addressed all your comments, let me know if that works with you !

ylacombe committed:
Merge branch 'huggingface:main' into add-w2v2-shaw

ylacombe committed:
improve converting script with processor class

amyeroberts reviewed (commented):
Thanks for the continued work on this! 

Just a few last things to address but otherwise looks good 🤗

ylacombe commented on utils/check_docstrings.py:
```diff
@@ -762,6 +762,7 @@
     "VitMatteForImageMatting",
     "VitsTokenizer",
     "VivitModel",
+    "Wav2Vec2BertForCTC",
```
The docstring check raises errors and running `make fix-copies` doesn't do anything!
Note that a bunch of `ForCTC` models are in there as well (Wav2Vec2ConformerForCTC, WavLMForCTC, HubertForCTC).

cc @ydshieh !

ylacombe committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ylacombe committed:
remove gumbel class

ylacombe committed:
remove tests related to previously deleted class

ylacombe committed:
Update src/transformers/models/wav2vec2_bert/configuration_wav2vec2_bert.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ylacombe commented on src/transformers/models/wav2vec2_bert/processing_wav2vec2_bert.py:
```diff
@@ -0,0 +1,142 @@
+
+        if audio is None and text is None:
+            raise ValueError("You need to specify either an `audio` or `text` input to process.")
+        elif text is not None and audio is not None:
+            raise ValueError(
+                "`Text` and `audio` are mututally exclusive when passed to `Wav2Vec2BertProcessor`. Specify one or another."
+            )
```
Currently, it's mutually exclusive, so not possible. I believe it remove the hassle of understanding which kwargs goes with which class. Note that the current usage seems to pad each modality independently (e.g [here](https://github.com/huggingface/transformers/blob/ff86bc364d20a4e6e3c0d954485eab959c803394/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py#L319-L331)).

Of course, open to removing this exclusivity and make it compatible to both modalities at the same time!

amyeroberts replied:
The reason I ask is whether we need a processor at all i.e. are there tasks which need both the feature extractor and tokenizer? It's fine if we need one for input and one for output - but if the input/output pairs only ever use one, then we shouldn't use a processor class

ylacombe replied:
I've updated the code to go back to Wav2Vec2Processor's original [behaviour](https://github.com/huggingface/transformers/blob/0cdcd7a2b319689d75ae4807cfb7b228aa322f83/src/transformers/models/wav2vec2/processing_wav2vec2.py#L128-L134). 

What bothers me is that the label attention mask is not used in the "forward" and "pad" methods when both modalities are passed. From the [`run_speech_recognition_ctc.py`](https://github.com/huggingface/transformers/blob/0cdcd7a2b319689d75ae4807cfb7b228aa322f83/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py#L333-L334), 

From the [CTC training example](https://github.com/huggingface/transformers/blob/0cdcd7a2b319689d75ae4807cfb7b228aa322f83/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py#L333-L334) and from looking at the `ForCTC` models, I reckon that labels are padded with `-100`. I can maybe adapt this behaviour, but anyways it is something that's supposed to be done in every fine-tuning examples. 

WDYT is the best way to solve this question? I'm personally in favor of staying with the current W2V2's behavior.

cc @sanchit-gandhi for visibility as well

amyeroberts replied:
> This still has the advantages of being consistent with previous W2V2 model, and having less import, easier from_pretrained/push_to_hub, and less overall complexity as compared to using both classes separately for the end-to-end user IMO.

Being consistent with other w2v2 models makes sense. I'm not sure I see the point about imports and from_pretrained though? If a model only uses a tokenizer, importing, saving and loading it is the same as using a processor. 

Processors are there to bundle together processing objects which are both needed for a single dataset -> inputs -> model -> outputs -> processed_outputs pass - not just for related objects. 

> What bothers me is that the label attention mask is not used in the "forward" and "pad" methods when both modalities are passed.

Could you expand on this further? What exactly is happening and what is the issue caused? 

I wouldn't worry too much about being compatible with training scripts - other than for backwards compatibility. The processor defines behaviour - the training scripts are then demonstrations of how to use them.

ylacombe replied:
>  If a model only uses a tokenizer, importing, saving and loading it is the same as using a processor.

The `ForCTC` models use both the tokenizer and feature extractor, so the processor import is easier than importing both the tokenizer and feature extractor here!

> Could you expand on this further? What exactly is happening and what is the issue caused?

I'm taking the example of the original Wav2Vec2 processor, since I've kept same behaviour here for now. 
https://github.com/huggingface/transformers/blob/07ae53e6e77ec6ff4fb25fbacfec4b11cfc82749/src/transformers/models/wav2vec2/processing_wav2vec2.py#L128-L134

As you can see, when both `audio` and `text` are passed, the `text` input_ids are used as a target, but the `text` attention mask is not passed through.
However, if you want to train the CTC model, you have to pad the input ids with a negative value (as shown [here](https://github.com/huggingface/transformers/blob/0cdcd7a2b319689d75ae4807cfb7b228aa322f83/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py#L333-L334) in the examples) because the `text` attention mask is infered in the CTC forward pass (as shown [here](https://github.com/huggingface/transformers/blob/0cdcd7a2b319689d75ae4807cfb7b228aa322f83/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L1988-L1990) in the modeling code).

Note that this CTC behaviour is common across all CTC classes. Also note that the W2V2 processor (and this new processor) are only meant to be used for CTC training.

To sum up, the current behaviour is to that the user himself pad the labels with a negative value with the `text` attention mask. To get it, the users currently do have to pass the audio and text separately. Ideally, the processors should take care of this but the already integrated processors don't.

amyeroberts replied:
If the current processors only process `text` or `audio`, then I don't think it's breaking anything to enable passing both inputs together and creating the attention mask as necessary - it's just extending what's there. 

The only thing I'd say is that the processor shouldn't be setting values to -100 to be ignored. This is a behaviour we always leave to the user.

ylacombe replied:
I've added `label_attention_mask` in [37dd941](https://github.com/huggingface/transformers/pull/28165/commits/37dd941d89a357b720028e8d4be1e0c4c9fd1079), let me know if that works for you. If necessary, I can revert to just adding "labels" when both modalities are passed.

ylacombe replied:
Hey @amyeroberts, following our offline discussions, I revert it to just adding `labels` when both modalities are passed.

ylacombe committed:
correct typos

ylacombe committed:
remove uused parameters

ylacombe commented on utils/not_doctested.txt:
```diff
@@ -873,6 +873,7 @@ src/transformers/models/wav2vec2/convert_wav2vec2_original_pytorch_checkpoint_to
 src/transformers/models/wav2vec2/convert_wav2vec2_original_s3prl_checkpoint_to_pytorch.py
 src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py
 src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py
+src/transformers/models/wav2vec2_bert/convert_wav2vec2_seamless_checkpoint.py
```
It imports `fairseq2`. Somewhat scripts to convert are always tested

ylacombe committed:
update processor to takes both text and audio

ylacombe committed:
update checkpoints

ylacombe committed:
update expected output and add ctc expected output

ylacombe committed:
add label_attention_mask

ylacombe committed:
replace pt with np in processor tests

amyeroberts commented on tests/models/wav2vec2_bert/test_processor_wav2vec2_bert.py:
```diff
@@ -1,7 +1,7 @@
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
+# you may not use this file excenp in compliance with the License.
```
I think there was a cheeky replace all here 😄

ylacombe replied:
Oups, nice catch

ylacombe replied:
Corrected by [7160906](https://github.com/huggingface/transformers/pull/28165/commits/71609062b4ca80e92f6606e02fb58de5978699fc)

ylacombe committed:
fix typo

ylacombe committed:
revert to behaviour with labels_attention_mask

amyeroberts reviewed (approved):
Thanks for all the work adding this and iterating on it!

ylacombe commented on src/transformers/models/wav2vec2_bert/modeling_wav2vec2_bert.py:
```diff
@@ -0,0 +1,1667 @@
+        )
+
+    # epsilon is used for probabilistic rounding
+    epsilon = np.random.rand(1).item()
```
I don't know tbh, but note that this method is using numpy only and that this is supposed to be run in preprocessing according to docs! It probably was copied out from the original repo
